### ISSUE TITLE
PostCSS plugins can't receive configuration options

### ISSUE THREAD

**Issue by jbalsas** [liferay/liferay-js-themes-toolkit#126]
From https://github.com/liferay/liferay-themes-sdk/issues/46#issuecomment-447167074

The current logic to generate the necessary postcss options [only accepts a list of plugin names](https://github.com/liferay/liferay-themes-sdk/blob/master/packages/liferay-theme-tasks/tasks/build/compile-css.js#L132-L141), which makes it imposible to pass down configuration to the plugins unless they specify another mechanism such as `package.json` extensions or its own `.rc` files.

We should allow this when configuring the postcss settings. One possibility would be:

```json
{
    "browserslist": ["last 4 version"],
    ...      
    "liferayTheme": {
        "baseTheme": "styled",
        "postcss": [
            "plugin1",
            {
                "name": "autoprefixer",
                "options": {
                    "cascade": false,
                    "remove": false
                },
            },
            "plugin3"
        ],
        "screenshot": "",
        "rubySass": false,
        "templateLanguage": "ftl",
        "version": "7.1"
    },
    ...
}
```

**Comment by yuchi** [comment]
I don‘t think this is a good approach to option passing.

The _Babel/Eslint/Browserify Way™_ is (as far as I remember) this:

```js
({
  plugins: [
    [ "plugin-name", { ...options } ],
    "plugin-with-no-options",
    [ "blah-blah" ],
  ]
})
```

**Comment by yuchi** [comment]
But looking [at PostCSS CLI](https://github.com/postcss/postcss-cli#context) you see the more idiomatic `{ [plugin]: options }` approach — but that looks like a “runtime options building” so probably having the configuration in the package.json is simply… wrong?

**Comment by jbalsas** [comment]
Hey @yuchi, that other syntax has always been a bit confusing to me, but I wouldn't mind using it that at all.

Why do you think we shouldn't have the configuration in the `package.json`? What would be the alternative? Another `.rc` file?

**Comment by jbalsas** [comment]
Now that I think about it... I think there's a `postcss.config.js` that might be available already... checking that

**Comment by jbalsas** [comment]
Yeah, so it looks like we should be able to use [postcss-load-config](https://www.npmjs.com/package/postcss-load-config), so you can simply write a `postcss.config.js` or a `.postcssrc.js` file to configure it.

This however, requires that we call `postcss` with no params, so we'll need to do some changes.

How does that sound?

**Comment by yuchi** [comment]
I honestly don’t now. That means that you cannot offer a default configuration, right? That could be cumbersome. Really, you got me off guard on this.

**Comment by jbalsas** [comment]
That might be a first, @yuchi! ;)

/cc @sbaechler, @protoEvangelion, mind offering your take on this?

**Comment by protoEvangelion** [comment]
@jbalsas You are headed in the right direction with using `postcss-load-config`. One of the postcss contributors addresses the issue about passing options to plugins [here](https://github.com/postcss/postcss-cli/issues/144#issuecomment-312913729).

It looks like the [_PostCSS CLI Way™_](https://github.com/postcss/postcss-cli#config) is to use `postcss.config.js` to accomplish this 😉 That being said I would be happy with any of the formats supported by `postcss-load-config`.

Regarding share-ability (_maybe this is beyond the scope of what we are talking about_), say I have two themes. Would it be feasible to expect that adding one `postcss.config.js` file at the root level could be used by both themes?

**Comment by gcmznt** [comment]
I everyone! I need to configure autoprefixer for a commerce theme and I am trying to figure out which way can be the best option, but I have not seen any cons with the solution that you have written here, from [this one](https://github.com/liferay/liferay-js-themes-toolkit/issues/126#issuecomment-448142674), to [this](https://github.com/liferay/liferay-js-themes-toolkit/issues/126#issuecomment-448151487). Just for instinct I prefer the `postcss-load-config` way of passing configuration to PostCSS

**Comment by wincent** [comment]
@gcmznt: just to make this more concrete, what is the configuration that you'd like to provide for autoprefixer? That way we have a very concrete example on which to base the discussion of the solution.

**Comment by gcmznt** [comment]
This is my actual configuration:
```
{
	grid: true,
	browsers: [
		"last 2 version",
		"> 1%",
		"IE 11"
	]
}
```

**Comment by wincent** [comment]
Didn't mean to close this yet. PR is not reviewed or merged yet. In meantime, I am going to make another PR for the 8.x branch.

**Comment by wincent** [comment]
Just published v8.0.6 and v9.0.0-beta.1 which contain the functionality for this. We went with the "configure using standard PostCSS configuration mechanisms" approach; that is, you can use "postcss.config.js", ".postcssrc" (etc... any of the mechanisms supported by [postcss-load-config](https://www.npmjs.com/package/postcss-load-config)). Note that you can still provide the simple name-only config via the "liferayTheme" -> "postcss" package.json properties, but it is an either-or thing (ie. you can only configure it one way, not both ways at the same time; if you try, we throw an error).

We've chosen not to document or advertise the functionality yet, until we're sure this is the right approach. Once we're convinced that this is right, we can go ahead and document it officially.

**Comment by gcmznt** [comment]
Thanks
